### PR TITLE
Cold users/items support in non-personilized models

### PR DESCRIPTION
**Pull request by feldlime**
## Description

- Supported cold recommendations in non-personalized models (random, popular and popular-in-category)
- Added check for incorrect types of cold targets
- Changed the logic of choosing random sampler for `RandomModel` and increased the sampling speed
- Changed the logic of `RandomModel`: now the recommendations are different for repeated calls of recommend methods

Closes https://github.com/MobileTeleSystems/RecTools/issues/119

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Optimization

## How Has This Been Tested?
Before submitting a PR, please check yourself against the following list. It would save us quite a lot of time.
- Have you read the contribution guide?
- Have you updated the relevant docstrings? We're using Numpy format, please double-check yourself
- Does your change require any new tests?
- Have you updated the changelog file?

### CODE REVIEW

File: rectools/models/random.py [hunk 1]
```diff
@@ -118,3 +118,27 @@ def _recommend_i2i(
         sorted_item_ids_to_recommend: tp.Optional[InternalIdsArray],
     ) -> tp.Tuple[InternalIds, InternalIds, Scores]:
         return self._recommend_u2i(target_ids, dataset, k, False, sorted_item_ids_to_recommend)
+
+    def _get_filtered_item_ids(self, sorted_item_ids_to_recommend: tp.Optional[InternalIdsArray]) -> InternalIdsArray:
+        if sorted_item_ids_to_recommend is not None:
+            item_ids = np.unique(sorted_item_ids_to_recommend)
```
Comment (blondered) [inline rectools/models/random.py:124]: `sorted_item_ids_to_recommend` are already after `np.unique`. Why do we do it here?

File: rectools/models/random.py [hunk 2]
```diff
@@ -22,17 +22,26 @@
...
+        self.values = values
+        self.values_list = list(values)  # for random.sample
+        self.python_gen = random.Random(random_state)
+        self.np_gen = np.random.default_rng(random_state)
+
+    def sample(self, n: int) -> np.ndarray:
+        if n < 25:  # Empiric value, for optimization
+            sampled = np.asarray(self.python_gen.sample(self.values_list, n))
+        else:
+            sampled = self.np_gen.choice(self.values, n, replace=False)  # pragma: no cover
```
Comment (blondered) [inline rectools/models/random.py:43]: Maybe add test here? Happy path at least?

Comment (blondered) [inline rectools/models/random.py:43]: xMaybe add test here? Happy path at least?

File: rectools/models/random.py [hunk 3]
```diff
@@ -71,18 +83,11 @@ def _recommend_u2i(
...
-            item_ids = np.unique(sorted_item_ids_to_recommend)
-        else:
-            item_ids = self.all_item_ids
-
-        item_indices = list(range(item_ids.size))  # for random.sample
-
-        np.random.seed(self.random_state)
-        random.seed(self.random_state, version=2)
+        item_ids = self._get_filtered_item_ids(sorted_item_ids_to_recommend)
+        sampler = _RandomSampler(item_ids, self.random_state)
```
Comment (blondered) [inline rectools/models/random.py:87]: We will have the same recommendations for multiple `recommend` calls. Is this our desired behaviour? I'm not sure.

File: rectools/models/random.py (hunk 2 above)
Comment (feldlime) [reply rectools/models/random.py:43]: duplicate?

File: rectools/models/random.py (hunk 3 above)
Comment (feldlime) [reply rectools/models/random.py:87]: It's a good point to think.

Actually we had the same behaviour before, so I haven't changed anything. 
But probably it'd be really better to change this.

File: rectools/models/random.py (hunk 1 above)
Comment (feldlime) [reply rectools/models/random.py]: Fixed

File: rectools/models/random.py (hunk 3 above)
Comment (feldlime) [reply rectools/models/random.py]: fixed

File: rectools/models/random.py (hunk 2 above)
Comment (feldlime) [reply rectools/models/random.py]: Added

File: rectools/models/random.py (hunk 3 above)
Comment (blondered) [reply rectools/models/random.py]: Probably should add this to Changelog

File: rectools/models/random.py [hunk 4]
```diff
@@ -118,3 +126,27 @@ def _recommend_i2i(
         sorted_item_ids_to_recommend: tp.Optional[InternalIdsArray],
     ) -> tp.Tuple[InternalIds, InternalIds, Scores]:
         return self._recommend_u2i(target_ids, dataset, k, False, sorted_item_ids_to_recommend)
+
+    def _get_filtered_item_ids(self, sorted_item_ids_to_recommend: tp.Optional[InternalIdsArray]) -> InternalIdsArray:
```
[the remark this answers is not in the document]
Comment (feldlime) [reply rectools/models/random.py]: removed

**Review by blondered — APPROVED** [review]

**Pull request merged**